modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -294,15 +314,20 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
                 pre_downsampling_size = target.shape[0]
 
                 with torch.autocast(self._device_type, enabled=self._amp):
-                    if self._downsampling_enabled:
+                    if self._downsampling_mode == DownsamplingMode.BATCH_THEN_SAMPLE:
                         # TODO(#218) Persist information on the sample IDs/weights when downsampling is performed
                         assert self._downsampler is not None
+                        self._downsampler.init_downsampler()
                         big_batch_output = self._model.model(data)
-                        downsampled_indexes, weights = self._downsampler.sample(big_batch_output, target)
-                        data, target, sample_ids = get_tensors_subset(downsampled_indexes, data, target, sample_ids)
+                        self._downsampler.inform_samples(sample_ids, big_batch_output, target)
+                        selected_indexes, weights = self._downsampler.select_points()
+                        selected_data, selected_target = get_tensors_subset(selected_indexes, data, target, sample_ids)
+
+                        sample_ids, data, target = selected_indexes, selected_data, selected_target
                         # TODO(#219) Investigate if we can avoid 2 forward passes
 
                     output = self._model.model(data)
+
                     if self._weighted_opt:
                         # weighted gradient descent
                         assert weights is not None
```

Comment:
so I think `_weighted_opt` does not respect weights from the selector yet, and also I am not sure where the weights would be coming from in that case. I think that is not completely handled yet